### PR TITLE
Implement source

### LINKED ISSUE #2: Source: initial configuration, read pages
### Feature description

As the title implies.

### PR DESCRIPTION
**Pull request by hariso**
### Description

Fixes #2 

### Checklist
- [x] Configurable poll interval
- [x] Handle the rounding down of `last_edited_time` in Notion
- [x] Extract rich content from pages
- [x] Handle grandchild blocks
- [ ] Verify the correctness of positions

### Quick checks:

- [x] There is no other [pull request](https://github.com/github.com/conduitio-labs/conduit-connector-notion/pulls) for the same update/change.
- [ ] I have written unit tests.
- [ ] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: README.md
Comment (samirketema) [inline README.md:11]: This is source-only at the moment right? If so, may be worth calling that out

File: docs/cdc.md
Comment (samirketema) [inline docs/cdc.md:41]: Conduit question - is it not able to access Conduit's store? (i.e. badger, postgres, etc)

File: source.go
Comment (samirketema) [inline source.go:199]: good to know there's pagination in the Notion API!

File: source.go
Comment (samirketema) [inline source.go:202]: they haven't been fetched at this point yet, right - just added to the fetch IDs slice?

Comment (samirketema) [reply source.go:202]: actually - I think I'm wrong. reading more closely

File: README.md
Comment (hariso) [reply README.md:11]: Good suggestion, I'll do that.

File: docs/cdc.md
Comment (hariso) [reply docs/cdc.md:41]: Nope, a connector has no access to it whatsoever.

File: source.go
Comment (samirketema) [reply source.go:202]: yeah - I misunderstood. we fetched the record, but soon will convert the block to an SDK record. this makes sense!

Comment (hariso) [reply source.go:202]: You're correct in that the pages haven't been fetched at this point, only the IDs. If you need to read more closely, then I need to add more comments.:)

**Review by samirketema — APPROVED** [review]
looks good to me! would be worth getting a review from the core conduit team as well, but that shouldn't block merging IMO.

File: source.go
Comment (hariso) [inline source.go:257]: Based on a call the team had today, I'll change this from JSON to just plain text.

File: go.mod
Comment (lovromazgon) [inline go.mod:6]: Umm, I this won't work with the latest version of Conduit, is this on purpose?

File: source.go
Comment (lovromazgon) [inline source.go:184]: For the first version this is.. well.. acceptable 😅 just know that since we are returning `ErrBackoffRetry` if there are no changes the SDK will already do some sleeping between calls to `Read`, so the poll interval is more like a "minimum delay between polls". Ideally we would have a background job that fetches these changes and feeds them to `Read` via a channel or something (the s3 connector does it that way).

File: source.go
Comment (lovromazgon) [inline source.go:135]: I don't see any rate limiting anywhere ([info in docs](https://developers.notion.com/reference/request-limits#rate-limits)), does the client handle this behind the scenes? Otherwise a request might return an error and cause the pipeline to stop. I think this is quite important to get right if we want the connector to be stable.

File: source.go
Comment (lovromazgon) [inline source.go:158]: We are losing the hierarchy info here, is this acceptable?

File: source.go
Comment (lovromazgon) [inline source.go:235]: Now, I know there's no other way to get a list of pages, but I have to call out the [docs](https://developers.notion.com/reference/post-search#limitations-of-search) which explicitly say:

> Search is not guaranteed to return everything, and the index may change as your integration is iterating through pages and databases resulting in unstable results.

This might be something to write into the known limitations.

File: source.go
Comment (lovromazgon) [inline source.go:269]: Thinking out loud - could using this time format ever be the source of an issue in the connector? It contains the timezone for example, I wonder how this affects us when the machine changes to DST, or if the record is processed on another machine. I'm not sure if this is a problem or not, but I'd rather use a UNIX timestamp if you ask me, just to be safe. As a side benefit - it would probably be faster to parse / encode.

File: source.go
Comment (lovromazgon) [reply source.go:235]: Given this info it seems that we should be prepared to fetch a block/page that doesn't exist and not fail because of that.

File: source.go
Comment (hariso) [reply source.go:269]: I think you're right about the Unix timestamp, I'll do that. Thanks for the suggestion!

File: go.mod
Comment (samirketema) [reply go.mod:6]: Yes - at the moment the platform is running v0.2.1. We need this connector (or even, the initial version of this connector) to be compatible with the platform, until we can prep the platform to upgrade to latest Conduit

File: source.go
Comment (hariso) [reply source.go:184]: I guess the goal we want to achieve, is that the `Read` method doesn't block?

Comment (hariso) [reply source.go:184]: > in the ideal world `Read` blocks until there's a record or the context gets cancelled

So returning an `ErrBackoffRetry` isn't actually recommended?

File: source.go
Comment (hariso) [reply source.go:257]: This will be handled in #7.

File: source.go
Comment (hariso) [reply source.go:135]: Lovro and I discussed this via Slack. While this is very important (any page which is slightly more complex will contain many blocks), we'll implement it in a follow-up PR (tracking issue #6), so that Samir is unblocked in his work.

File: source.go
Comment (hariso) [reply source.go:235]: Correct... And even if the search would be reliable, a page can be deleted between when we got its ID and when we attempted to fetch it. Thanks for calling it out, I'll implement this in a follow-up PR.

Comment (hariso) [reply source.go:235]: Tracked here: #8.

File: source.go
Comment (hariso) [reply source.go:158]: I believe so, because we'll "flatten" a page into a plain text file.

Comment (hariso) [reply source.go:159]: I'll resolve this conversation to be able to merge the PR, but I'll be happy to any implement fixes/improvements here in a follow-up PR.

File: source.go
Comment (hariso) [reply source.go:185]: I'll resolve this conversation to be able to merge the PR, but I'll be happy to any implement fixes/improvements here in a follow-up PR.

**Review by lovromazgon — APPROVED** [review]

File: source.go
Comment (lovromazgon) [reply source.go:185]: Regarding `ErrBackoffRetry` I'd say it's just a utility that makes your life easier. It allows you to easily get started with polling, but you (the developer) are not in charge of the delay between the retries, so it's not ideal (well we have an [issue](https://github.com/ConduitIO/conduit/issues/184) for that). It's not _wrong_ to use it, but not the best either.

You also have to consider that there might be significant delays between calls to `Read` if the pipeline is "full" (I'm talking about backpressure in pipelines with slow destinations). In that case, we are losing precious time when we could be already downloading the next page and preparing it so that when `Read` is called we already have a record ready to return. So having a background job that downloads records and stores them in a cache makes the code a bit more complicated, but in turn it keeps the connector completely in charge of the polling interval and makes it more performant by downloading pages even while the SDK is busy returning a record back to Conduit.

Comment (hariso) [reply source.go:185]: FTR: we'll implement it in a follow-up PR (tracking issue https://github.com/conduitio-labs/conduit-connector-notion/issues/12), so that Samir is unblocked in his work.

**Pull request merged**
